mysqlnd_azure.c

```diff
@@ -447,6 +506,9 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
                 }
 
                 if (MYSQLND_AZURE_G(enableRedirect) == REDIRECT_PREFERRED) {
+                  if (MYSQLND_AZURE_G(logLevel > 1)) {
+                    AZURE_LOG("INFO ", "Redirection conn connect failed, mysqlnd_azure.enableRedirect = PREFERRED, fall back to classical connection.");
```

Comment:
typo. And maybe we can specify the error reason due to handshake here

Reply:
And for the else cluase, //mysqlnd_azure.enableRedirect=on and redirect connection failed, there is a corresponding error message, the original established connection is closed here, and it will use MySQL's error message as return message. Maybe we should log a ERROR line here to indiate that the original established connection is closed since mysqlnd_azure.enableRedirect=on

Reply:
But you have moved the redirect_conn context to conn after conn resources released.

> Maybe we should log a ERROR line here to indiate that the original established connection is closed since mysqlnd_azure.enableRedirect=on
 Why should we log this?

Besides, I wonder why we copy the redirect_conn context back to conn?

Reply:
I think a log here may be good since redirect_handshake failed here, and because enableRedirect=on, the original conn is also closed here. That's the reason why no connection is avaible. But since it will finally return redirect_handshake's error, no log here should also be fine.

The reason I copy context back to pconn is that I want to use the error section (below in this method) and the up level to handle the error for me, and error is stored in redirect_conn object, pconn is also a parent object, it will bring the error conn object back to uplevel, and report error correspondly.   Copy back to pfc is actually not necessary in error case since this pointer will only be used below in after_conn section. You may ignore it